rootfs/etc/nginx/lua/test/balancer_test.lua

```diff
@@ -1,3 +1,299 @@
+end
+
+local function backend_generator(name, endpoints, lb_alg)
+  return {
+    name = name,
+    endpoints = endpoints,
+    ["load-balance"] = lb_alg,
+  }
+end
+
+local default_endpoints = {
+  {address = "000.000.000", port = "8080"},
+  {address = "000.000.001", port = "8081"},
+}
+
+local default_backends = {
+  mock_rr_backend = backend_generator("mock_rr_backend", default_endpoints, "round_robin"),
+  mock_ewma_backend = backend_generator("mock_ewma_backend", default_endpoints, "ewma"),
+}
+
+local function init()
+  mock_cjson = {}
+  mock_config = {}
+  mock_sticky = {}
+  mock_ngx_balancer = {}
+  mock_ewma = {}
+  mock_backends = dict_generator(default_backends)
+  mock_lrucache = {
+    new = function () return mock_backends end
+  }
+  lock = {
+    lock = function() return end,
+    unlock = function() return end
+  }
+  mock_lock = {
+    new = function () return lock end
+  }
+  _G.ngx = {
+    shared = {
+      round_robin_state = dict_generator({}),
+      balancer_ewma = dict_generator({}),
+      balancer_ewma_last_touched_at = dict_generator({}),
+    },
+    var = {},
+    log = function() return end,
+    WARN = "warn",
+    INFO = "info",
+    ERR = "err"
+  }
+  package.loaded["ngx.balancer"] = mock_ngx_balancer
+  package.loaded["resty.lrucache"] = mock_lrucache
+  package.loaded["cjson"] = mock_cjson
+  package.loaded["resty.lock"] = mock_lock
+  package.loaded["balancer.ewma"] = mock_ewma
+  package.loaded["configuration"] = mock_config
+  package.loaded["sticky"] = mock_sticky
+  balancer = require("balancer")
+end
+
 describe("[balancer_test]", function()
-  -- TODO: write tests 
+  setup(function()
+    init()
+  end)
+
+  teardown(function()
+    local packages = {"ngx.balancer", "resty.lrucache","cjson", "resty.lock", "balancer.ewma","configuration", "sticky"}
+    for i, package_name in ipairs(packages) do
+      package.loaded[package_name] = nil
+    end
+  end)
+
+  describe("balancer.call():", function()
+    setup(function()
+      mock_ngx_balancer.set_more_tries = function () return end
+      mock_ngx_balancer.set_current_peer = function () return end
+      mock_ewma.after_balance = function () return end
+    end)
+
+    before_each(function()
+      _G.ngx.get_phase = nil
+      _G.ngx.shared.round_robin_state._vals = {}
+      _G.ngx.var = {}
+      mock_backends._vals = default_backends
+      mock_sticky.is_sticky = function(b) return false end
+    end)
+
+    describe("phase=log", function()
+      before_each(function()
+        _G.ngx.get_phase = function() return "log" end
+      end)
+
+      it("lb_alg=ewma, ewma_after_balance was called", function()
+        _G.ngx.var.proxy_upstream_name = "mock_ewma_backend"
+
+        local backend_get_spy = spy.on(mock_backends, "get")
+        local ewma_after_balance_spy = spy.on(mock_ewma, "after_balance")
+
+        assert.has_no_errors(balancer.call)
+        assert.spy(backend_get_spy).was_called_with(match.is_table(), "mock_ewma_backend")
```

Comment:
This does not add any value IMHO and is leaking internal knowledge about implementation.